design/ceph/osd-migration.md

```diff
@@ -0,0 +1,105 @@
+
+#### OSD Prepare
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
+    - RAW MODE
+      - `ceph-volume raw prepare --bluestore_rdr --data /dev/vda`
+    - LWM MODE
+      - `ceph-volume lvm prepare --bluestore_rdr --data /dev/vda`
+
+#### OSD Create
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+  - **OSD on disks**:
+    - RAW MODE
+    `ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs`
+
+    - LVM MODE
+    `ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"`
+
+
+
+### EXISTING CLUSTERS
+- After upgrading an existing rook ceph cluster to a version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+  - Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+- Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label does not match the osd store type provided in `spec.storage.osdStore.type`. Lets call it `OSDDeleteList`.
+  2. If the PG status is not `active+clean`, cancel the osd replace operation and reconcile the operator.
+  3. Once the PG status is `active+clean`, delete a single OSD from the `OSDDeleteList`. Refer the deletion process mentioned below.
+  4. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDDeleteList`. Skipping update operation on the OSDs that need to be migrated is important because with OSDs on disks, the OSD prepared pod is updated if there is an update in the spec. This will update the osd store to `bluestore-rdr` in the spec without migrating the OSD. This will cause errors.
+  5. Once the OSD is added back, reconcile the operator again to continue replacing the other OSDs by following the same steps mentioned above.
+
+
+
+#### DELETE OSD
+OSD can be deleted by using the following options:
+
+##### OSD PURGE JOB
+- Stop the OSD pod by patching the OSD deployment with `replica=0`
+- Mark the OSD as `down`
+- Once the OSD is `down`, run the OSD purge job to forcefully delete the OSD. No need to check if the OSD `safe-to-destroy`
```

Comment:
Still need to update this section since no purge job is used, right?

Reply:
I've update the design doc.

Reply:
marking as resolved.